# GCB/CIS/MTR 999 (Spring 2022): Introduction to Bioinformatics

## Course Staff

### Course Directors

Please use e-mail to contact the course instructors for administrative questions.
For technical questions about materials, please contact via Slack.

#### Benjamin F. Voight, PhD
    Department of Systems Pharmacology and Translational Therapeutics
    Department of Genetics
    10-126 Smilow Center for Translational Research (SCTR)
    3400 Civic Center Blvd
    215-746-8083
    bvoight@pennmedicine.upenn.edu

#### Ophir Shalem, PhD
    Department of Genetics
    Center of Cellular and Molecular Therapeutics
    3501 Civic Center Boulevard
    5th floor CTRB, room 5050
    shalemo@upenn.edu


### Keywords

* Ben Voight: human genetics, statistics, cardiometabolic disease
* Ophir Shalem: functional genomics, protein quality control, neurodegenerative disease

## Location, time, and contact

This independent study will self-directed for Spring 2022.

We will schedule a regular "check-in" time, where professors will be available to troublshoot or answer question in real time.

That will be planned subsequently during the first meeting of the independent study.

The first meeting on the class is planned (tentatively) for:

    Wednesday, Jan 12th, 2022
    11a - 12p, ET

We plan to have a slack channel from which communication between students and with Ophir and Ben is available.

## Course web site

We will use [CoCalc](https://cocalc.com/) for in class exercises and homework.

## Course description

This independent study (GCB999) mirrors the offering of GCB/CIS/MTR535, Introduction to Bioinformatics.

This course provides broad overview of bioinformatics and computational biology
as applied to biomedical research. A primary objective of the course is to
enable students to integrate modern bioinformatics tools into their research
activities. Course material is aimed to address biological questions using
computational approaches and the analysis of data.

A basic primer in programming and operating in a UNIX environment will be
presented, and students will also be introduced to Python, R, and tools for
reproducible research. This course emphasizes direct, hands-on experience with
applications to current biological research problems. The course is **NOT
INTENDED** for computer science students who want to learn about biologically
motivated algorithmic problems; GCB/CIS/BIOL536 and GCB537 is more appropriate.

Areas include DNA sequence alignment, genetic variation and analysis, motif
discovery, study design for high-throughput sequencing, RNA and gene expression,
single gene and whole-genome analysis, machine learning, and topics in systems
biology. The relevant principles underlying methods used for analysis in these
areas will be introduced and discussed at a level appropriate for biologists
without a background in computer science.

The course will assume a solid knowledge of modern biology. An advanced
undergraduate course such as BIOL421 or a graduate course in biology such as
BIOL526 (Experimental Principles in Cell and Molecular Biology), BIOL527
(Advanced Molecular Biology and Genetics), BIOL528 (Advanced Molecular
Genetics), BIOL540 (Genetic Systems), or equivalent, is a prerequisite.

## Note on Reproducibility and Responsible Research

Scientific rigor and reproducibility (SRR) is essential for progress, and especially so in 
Bioinformatics. This course aims to address SRR by teaching students programming
and computational pipeline building skills, so that students can generate results that are
"technically reproducible" - that the same code run on the same input file generates
the same output file or result each time. Moreover, the class will teach students
technical an intuitional skills on how to look at, and evaluate, data sets they are 
provided that might be established or new. This helps to generate SRR because 'bugs'
or technical artefacts in data files hinders the production of reproducible results. 
Students who leave the course will have developed their intuition and 'healthy skepticism'
new data sets they encounter, and a framework to evaluate, visualize, and correct
issues in data as they are encountered.

## Equipment prerequisite

All students will need a laptop for activities.

## Schedule

The most up to date schedule for the course is available in the [course
calendar](https://github.com/bvoight/GCB535/blob/master/Course_Calendar.md).

## Grading

Given the fluid nature of this independent offering, we ask students to complete
work on an ~monthly basis, but we will provide at least 1 months worth of materials
in advance such that student could completely some/all of the work far ahead of time.

As a result, there will not be a "late" assignment policy. Grades will be based on the 
number of modules that are completed, which includes the 'active exercise' and 'homework' 
problems provided. 

We will drop scores from the 3 lowest scoring assignments when determining the final grade.

All modules that count for the final grade must be completed by April 27th to be considered. 

## Plagiarism policy

Consistent with the University of Pennsylvania's honor code and policies on
academic integrity, we maintain a zero-tolerance policy on plagiarism. For
assignments containing text that the instructors determine is plagiarized,
students will receive a grade of zero for the assignment and the case will be
referred to the Office of Student Conduct for disciplinary measures. Students
may not always be aware of what constitutes plagiarism in their work. If you are
unsure, please contact one of the course directors. Please see links below
regarding the office of academic integrity for Penn’s policy on plagiarism and
our discretion on grading.

https://provost.upenn.edu/policies/faculty-handbook/students/iv-d

http://www.upenn.edu/academicintegrity/ai_violations.html
