{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Genomic Databases - In Class Activity"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Imagine that you are a grad student who has just begun work in a mouse lab. Your advisor, Dr. Stoker, works on a novel mouse phenotype, which has been dubbed Vampiric. Mice with this phenotype have the physiological feature of exceptionally sharp teeth and the behavioral feature of biting other mice. Dr. Stoker has developed a genetic mutagenesis screen for this phenotype using sunlight as a negative selector. In one strain of Vampiric mice, the mutation has been narrowed to an approximately 250-kilobase (kb) region of chromosome 12. Your first task is to investigate what is known about candidate genes in this region."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Part I. Explore the basic functionality of the UCSC Genome Browser"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1. Go to http://genome.ucsc.edu in your favorite browser. Take a moment to read `Our History` from the `About Us` drop down menu.\n",
    "\n",
    "2. Then, check out `News` from the `About Us` drop down menu. You'll see that this database is pretty active, and is constantly and rapidly providing updates!\n",
    "\n",
    "3. Click the link at the top left that reads `Genomes` to access genomes of your choice. Your queries should go in the text box under `Position/Search term`. Read through the `Sample Position Queries` further down the page for an explanation of how to search for a particular region.\n",
    "\n",
    "***Be mindful! Genomes have versions and it is important that you know which version of the genome you want to query, or you may not find what you want (or get 'where' appropriately annotated)!***\n",
    "\n",
    "3. Choose a different species from the `genome` drop-down list. Note that the entire page changes."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Part II. Investigate the genes in Dr. Stoker’s candidate region"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1.\tSearch for the region on: chromosome 12 between bases 56,532,042 and 56,785,902 on the previous assembly of the Mouse genome (`GRCm38/mm10`). Choose `Mouse GRCm38/mm10` from the `genome` drop-down list.\n",
    "\n",
    "Type or copy-paste `chr12:56532042-56785902` in the `search term` cell. Click `go`.\n",
    "\n",
    "2. You will be taken to a page with an image displaying tracks of annotation information for this region. Take a moment to explore this page:\n",
    "\n",
    "* Controls for shifting the display region or zooming in/out are above the track image. \n",
    "* Some tracks can be expanded or compressed by clicking on them. In other cases, individual annotations on tracks (such as genes) can be clicked for more details. \n",
    "* You can also right click on the right most side of the track (the 'Y-axis'), which can also change the form of viewing the track, but also will give you an option to change the *preferences* for the track too. Depending on the track, this can give you different features, annotations, or functionality. You can also access these preferences by click on on the `configure` button.\n",
    "\n",
    "* Below the track image is a list of the available tracks for the genome assembly that you are viewing. You can control which tracks are shown by changing the drop-down boxes next to each track name and then clicking “refresh” (the button on the page, not on the browser toolbar). Note that these may depend on the genome version you are looking at! (for example, the most recent genome build, `GRCm39/mm39`, does not have a track annotation for the single-cell profiling experiments performed in Mouse generated by the [Tabula Muris](https://tabula-muris.ds.czbiohub.org/) project, published in [2018](https://www.nature.com/articles/s41586-018-0590-4)).\n",
    "\n",
    "* You can also guarantee that you are looking at a standard set of tracks by clicking the menu item `view` and then `default tracks`. In the items below, you'll be looking at a range of default tracks which we modify for visualization purposes."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To answer the following 3 questions, scroll down to the `Genes and Gene Prediction Tracks` and set \n",
    "- `NCBI RefSeq` to `pack`\n",
    "- `GenScan Genes` to `dense`\n",
    "\n",
    "Next, scroll down to `mRNA and EST Tracks` and set\n",
    "- `Mouse mRNAs` to `pack`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Q1.\tWhat RefSeq Mouse Genes are in this region?"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "RefSeq genes are a good place to start your search for candidate genes; however, it is possible that the mutation for Vampiric is in an unknown gene. Note that the Mouse mRNA track has many more annotations than the RefSeq track. RefSeq is a manually curated non-redundant gene database, while GenBank is a larger database with potentially redundant experimental data. For each gene in RefSeq there is typically one or more corresponding mRNA in GenBank that aligns to the same position in the genome. However, some mRNAs in GenBank may be unconfirmed as real genes, and therefore do not have a corresponding RefSeq entry. Such mRNAs would be further candidates for your mutation.\n",
    "\n",
    "RefSeq is a broader effort for annotation of genomic data -- extending beyond mouse and humans. There are more systematic efforts to more deeply annotate human and mouse data: [GENCODE](https://www.gencodegenes.org/), which is also provided as a default track on the browswer. You'll see that RefSeq and GENCODE generally agree, but [GENCODE carries extensive annotations](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9825462/) (for instance, in annotation of alternative splicing) and is integrated with Ensembl, a widely-utilized annotation informatics database."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Q2.\tBelow, list an example of:\n",
    "\n",
    "- an `GeneBank mRNA` in this region that does not correspond to a `NCBI RefSeq gene`. \n",
    "- a `GENCODE gene` in this region that does not correspond to `Genbank mRNA` or a `NCBI RefSeq gene`."
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You should also consider the possibility that your mutation is in a gene that has never been characterized experimentally. GenScan is a computational tool for predicting genes. Take a look at the annotation track for genes predicted by GenScan. You should always take these predictions with a grain of salt, but they may be useful if you do not see obvious results from known genes or mRNAs."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Q3.\tHow many GenScan predicted genes are in this region?"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Part III. Obtain more detailed information about one of the candidate genes, Pax9."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now it’s time to look at what is known about our candidate genes. Both `GENCODE` or the `NCBI RefSeq` will have useful information as they are often based on multiple experiments and have been validated in some way. \n",
    "\n",
    "Let's zoom in on the region around Pax9:\n",
    "- Head to the very top track where \"chr12\" is listed. click just upstream of the tick \"56,700,000\" before the Pax9 5' UTR, and draw a box around the region you want to zoom in on. Click `Zoom In` on the dialog box. (Alternatively, you can requery: `chr12:56687409-56715758`).\n",
    "\n",
    "The intron/exon structure of this gene is shown in the RefSeq track. The thickest lines represent exons, the medium lines represent untranslated regions, and the thin lines represent introns, with arrows indicating the direction of transcription.\n",
    "\n",
    "Click on Pax9 in the `NCBI RefSeq` track on the browser."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Accession numbers that begin with `NM` are for mRNA sequence entries, while accession numbers begun with `NP` are protein sequence entries. GenBank accession numbers, such as the ones seen in question 2, are for experimental data entries.\n",
    " \n",
    "Q4.1. What is the mRNA NCBI RefSeq Accession Number for this Pax9 annotated transcript?"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Q4.2. What is the *span* of this annotation for Pax9; i.e., the number of base pairs in the entire transcribed mRNA, including introns and untranslated regions?"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Q5. Is Pax9 on the forward or reverse strand of chromosome 12?"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Click on the different GENCODE transcripts for Pax9.\n",
    "\n",
    "Q6. What are the major differences among the two GENCODE and NCBI RefSeq annotated transcripts?"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Q7.\tHow many exons are in Pax9 (NCBI RefSeq annotation)? Does this change for the GENCODE annotations?"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Click back on the `GENCODE` and the first track for Pax9 (i.e., GENCODE transcript ID `ENSMUST00000001538.9`), under `Sequence and Links to Tools and Databases`, click `Entrez Gene`. This brings you to an NCBI page with more detailed information on this gene. The NCBI Entrez Gene database is linked to GenBank, PubMed, and many other useful databases. \n",
    "\n",
    "Scroll down to the section marked `Genomic regions, transcripts, and products`. You will see an image composed with a green colored line for genomic information, purple for mRNA information and red for protein information. If you only see green lines (genes), click on them and the green line will expand to purple line and red line. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Q8.\tConsider the RefSeq transcript we examined in the UCSC Browser, `NM_011041.3`. \n",
    "\n",
    "* Mouse over the purple line. What is the length of this Pax9 transcript (excluding introns)? \n",
    "* Mouse over the red line. How many amino acids are there in Pax9 protein? \n",
    "* Do they follow the 3:1 ratio (as 3 mRNA nucleotides code for 1 amino acid)? What is the biological reason for this discrepancy?"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If you hover over the purple line, you will see many options to display the genomic, mRNA, and protein information of Pax9 in different formats. If you select `FASTA record`, you'll see the gene sequence corresponding to Pax9. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Continue scrolling the NCBI gene page and investigating Pax9.\n",
    "Note that the `Summary` section, `GeneRIF` (in the `Bibliography` section), Gene Ontology (in the `General Gene Information` section), and `Interactions` section each provide useful information on what is known about the function of this gene. Use information contained in `GeneRIF`, `Gene Ontology`, and `Interactions` for the following:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Q9.\tWhat evidence (if any) supports Pax9 as a likely mutant related to the Vampiric phenotype?"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Q10. What other basic biological function (if any) does Pax9 have?"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Q11. What genes or proteins (if any) does Pax9 interact with?"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Part IV. Get information on Pax9 in other (non-mouse) species."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Imagine that you are able to confirm that Pax9 is in fact responsible for the Vampiric phenotype in Mouse. Now Dr. Stoker wants to try inducing this phenotype in other organisms. He asks you to find out which species have Pax9 genes in GenBank.\n",
    "1.\tGo to: http://www.ncbi.nlm.nih.gov/Genbank/\n",
    "Read the introductory information for GenBank. One common way to search GenBank is to submit a sequence as a BLAST search. However, we will cover BLAST and other sequence homology tools later on in the class. In the meantime, we will search the Gene database using keywords.\n",
    "2.\tGo to: http://www.ncbi.nlm.nih.gov/gene\n",
    "Read the “Help” section for extra hints on ways you can search.\n",
    "3.\tSearch for “Pax9” in the search bar at the top of the page.\n",
    "You will notice that you get a list of many genes but not all of them are called Pax9. This is because you have searched all fields in the database and are also seeing genes related to Pax9. Narrow your search by clicking “Advanced” search and modifying the Builder to “Gene Name”. You can also accomplish the same thing by changing your search string to: Pax9[sym], 'sym’ stands for the gene symbol. Species are typically listed on the first line of the item summary in brackets (e.g., [Gallus gallus])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Dr. Stoker suggests trying to induce the Vampiric phenotype in human subjects but you point out that that would be against ethical guidelines. He agrees and suggests instead that you use the species Gallus gallus (Chicken)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Q12.\tWhat chromosome is Pax9 on in Chicken? "
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Q13.\tIf you wanted to go back and view this region of the chicken genome in UCSC Genome Browser, what search string would you use?"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Part V. \"Liftover\" to different genome builds"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As mentioned above, there are often multiple builds of genomes, some of which are more or less (or sometimes, differently) annotated. In some cases, you may need to obtain the genomic coordinates that you have mapped from one build to a different build, i.e. \"Lift over\" annotations from one build to another\n",
    "\n",
    "For small-scale lookups, there exists a tool called `liftover` which can accomplish this for you. (NOTE: For LARGE-SCALE (data files >500 Mb) liftover activities -- for example, genome-wide), a UNIX command-line LiftOver tool is available. We will be learning more about UNIX later in the course!)\n",
    "\n",
    "Q14. For now, try lifting over our previous region:\n",
    "\n",
    "- Select from the menu item `Tools`, `Liftover`.\n",
    "- Select `Mouse`, `GRCm38/mm10` as the Original assembly\n",
    "- Select `Mouse`, and `GRCm39/mm39` as the new assembly to lift over onto\n",
    "- Provide the coordinates of the region (either as `chr12:56532042-56785902` or in [BED format](https://genome.ucsc.edu/FAQ/FAQformat.html#format1)).  \n",
    "\n",
    "What are the new coordinates for this region? How much did this region 'shift' in the new mapping?"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  },
  "name": "Online_database_in_class.ipynb"
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
