{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Bringing it all together: cross referencing data files\n",
    "---\n",
    "\n",
    "To wrap things up, we'll go over one big example that integrates a little bit of everything we've learned so far. This example will demonstrate a very common task in genomics: cross referencing data between different text files.\n",
    "\n",
    "***The situation:***\n",
    "\n",
    "I'm interested in looking at the distribution of translation start sites (TSS) along the length of mRNAs. I already have a file with the locations of all the TSS in mouse (`all_start_sites.txt`). However, first I need to normalize each TSS position by gene length so that I can graph the *relative* position of TIS across all genes in my dataset. I can easily do this by dividing the start site position by the full length of the gene. Unfortunately, though, I have gene lengths stored in a separate file (`transc_lengths.txt`), so I need to somehow integrate these two pieces of information (TSS position and gene length) across the two files.\n",
    "\n",
    "Here's a picture version of what I want to do:\n",
    "\n",
    "<img align=\"left\" src=\"tis_normalization.PNG\" />"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And here a few lines from each of the two files I have:\n",
    "\n",
    "*`all_start_sites.txt`*\n",
    "```\n",
    "knownGene\tGeneName\tInitCodon[nt]\tDisttoCDS[codons]\tFramevsCDS\tInitContext[-3to+4]\tCDSLength[codons]\tHarrPeakStart\tHarrPeakWidth\t#HarrReads\tPeakScore\tCodon\tProduct\n",
    "uc007afd.1\tMrpl15\t248\t79\t1\tAATATGG\t15\t247\t2\t368\t2.61\taug\tinternal-out-of-frame\n",
    "uc007afh.1\tLypla1\t36\t5\t0\tAACATGT\t225\t34\t4\t783\t3.27\taug\tn-term-trunc\n",
    "uc007afi.1\tTcea1\t28\t-24\t0\tGGCTTGT\t325\t27\t3\t446\t1.43\tnearcog\tn-term-ext\n",
    "uc007afi.1\tTcea1\t100\t0\t0\tGCCATGG\t301\t99\t3\t3852\t3.79\taug\tcanonical\n",
    "uc007afn.1\tAtp6v1h\t100\t-13\t-1\tGCTATCC\t10\t99\t3\t728\t0.77\tnearcog\tuorf\n",
    "uc007afn.1\tAtp6v1h\t149\t3\t0\tAAGATGG\t480\t147\t3\t1407\t1.36\taug\tn-term-trunc\n",
    "uc007agb.1\tPcmtd1\t120\t-97\t-1\tGCGCTGG\t45\t119\t3\t65\t0.75\tnearcog\tuorf\n",
    "...\n",
    "```\n",
    "\n",
    "*`transc_lengths.txt`*\n",
    "```\n",
    "SeqID\tLen\n",
    "uc009gmc.1\t4900\n",
    "uc008mue.1\t459\n",
    "uc007hzr.1\t4578\n",
    "uc007gtm.1\t1257\n",
    "uc007axo.1\t2311\n",
    "uc007wps.1\t2694\n",
    "uc007gqc.1\t30\n",
    "uc009smc.1\t1530\n",
    "...\n",
    "```\n",
    "\n",
    "The common piece of information between these files is the gene ID (`knownGene` and `SeqID`), so this is what we will use to match up start sites to gene lengths. Note that we can not count on the gene IDs to be in the same order in the two files, so we will need some other way of matching up the data. (Something like a lookup table, perhaps..)\n",
    "\n",
    "***The plan***\n",
    "\n",
    "When working with more than one file, it sometimes helps to write down a step-by-step plan before you start coding. One way to do this is to write out your plan in the form of code comments -- then, once you're ready to code, you can just fill in what's needed under each comment.\n",
    "\n",
    "Here's my step-by-step plan for tackling this problem:\n",
    "\n",
    "```\n",
    "1. Open length file\n",
    "2. For each line in length file: \n",
    "    a. Extract the gene id (1st column) and length (2nd column)\n",
    "    b. Store lengths in dictionary based on gene id (where gene IDs are the keys and lengths are the values)\n",
    "3. Open output file\n",
    "4. Open TSS file\n",
    "5. For each line in tss file:\n",
    "    a. Extract the id (1st column) and start site (3rd column)\n",
    "    b. Using the id, lookup the length of the gene from the hash\n",
    "    c. Divide the start position by the length of the gene\n",
    "    d. Print the result to the output file\n",
    "```\n",
    "\n",
    "***The code***\n",
    "\n",
    "And here's the final code. (Side note: I heavily commented this code for educational purposes, but usually you would not need so many comments!)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# input files\n",
    "tssFile = \"all_start_sites.txt\"\n",
    "lenFile = \"transc_lengths.txt\"\n",
    "\n",
    "# output files\n",
    "normOut = \"normalized_tss.txt\"\n",
    "\n",
    "# lookup table (dictionary) for transc lengths\n",
    "lengthDict = {}\n",
    "\n",
    "\n",
    "# read in lengths, store in dictionary\n",
    "ins = open(lenFile, 'r')\n",
    "ins.readline()                              # don't forget to skip the header!\n",
    "\n",
    "for line in ins:\n",
    "    line = line.rstrip('\\r\\n')\n",
    "    \n",
    "    linePartsList = line.split()            # split on whitespace to break line into separate pieces of data\n",
    "    geneID = linePartsList[0]               # gene id is the first column, so it's stored in index 0 of linePartsList\n",
    "    geneLength = int(linePartsList[1])      # gene length = second column, so it's stored in index 1. Note conversion to int!!\n",
    "    \n",
    "    lengthDict[geneID] = geneLength         # store len in hash labeled by the id\n",
    "\n",
    "ins.close()\n",
    "\n",
    "\n",
    "# read in TSS, use stored lengths to normalize, then print. No need to store TSS.\n",
    "skipCount = 0 \n",
    "\n",
    "output = open(normOut, 'w')\n",
    "output.write(\"RelativePos\\n\")               # write header line for output file. remember to add a \\n character to the end.\n",
    "\n",
    "ins = open(tssFile, 'r')\n",
    "ins.readline()                              # skip header in input file\n",
    "\n",
    "for line in ins:\n",
    "    line = line.rstrip('\\r\\n')\n",
    "    \n",
    "    linePartsList = line.split()            # split on whitespace to break line into separate pieces of data\n",
    "    geneID = linePartsList[0]               # gene id is the first column, so it's the first element in the list\n",
    "    tss = int(linePartsList[2])             # TSS = third column, so it's third in the list. Note conversion to int!!\n",
    "    \n",
    "    if geneID in lengthDict:                # make sure there is an entry in lengthDict for this id. See note below (##)\n",
    "        fullLen = lengthDict[geneID]        # using the gene id as the key, lookup the length of this gene in the dictionary\n",
    "        norm = tss / float(fullLen)         # get relative position of TSS. We convert one number to float to avoid int division\n",
    "        output.write(str(norm) + \"\\n\")      # write the result to output file. Note conversion to string and \\n on end.\n",
    "    else:\n",
    "        print \">Warning: geneID\", geneID, \"was not in length lookup table. This gene will be skipped.\"\n",
    "        skipCount = skipCount + 1\n",
    "\n",
    "ins.close()\n",
    "output.close()\n",
    "\n",
    "print \"Skipped\", skipCount, \"entries in TSS file.\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**(##) Note: Why do we have the line \"`if geneID in lengthDict:`\"?** \n",
    "\n",
    "If there are gene IDs in the TSS file that were not included in the length file, then we will get an error when we try to look up that ID in the lengthDict (because it's just not there). As you can see from the output above, this is exactly what would have happened to us had we not included this `if` statement (try removing it yourself and see what happens).\n",
    "\n",
    "Whenever you think there is a chance this sort of thing might happen (and there almost always is in genomics, unfortunately) just add a quick `if` statement like the one here, which will let you skip over any IDs that would cause an error. \n",
    "\n",
    "It's also usually a good idea to print out a message whenever you skip data, like we do in the `else:` statement above. However, this can sometimes cause your code output to get out of hand (as it does here), so sometimes it's better to just print the count of how many data points were skipped.\n",
    "\n",
    "***Code output***\n",
    "\n",
    "The main output of this code is a new file called `normalized_tss.txt` that contains all of the normalized TSS positions in a single column. Here's what it looks like:\n",
    "\n",
    "```\n",
    "RelativePos\n",
    "0.0147118921128\n",
    "0.0506072874494\n",
    "0.0754048582996\n",
    "0.0229226361032\n",
    "0.0506208213945\n",
    "0.0787010506208\n",
    "0.140783190067\n",
    "0.170009551098\n",
    "0.0329929300864\n",
    "...\n",
    "```\n",
    "\n",
    "These data can then be plotted in the form of a histogram. We won't go over plotting here, but here's an example of a plot generated using ggplot2 in R:\n",
    "\n",
    "<img align=\"left\" src=\"tss_dist.PNG\" />"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "From this we can conclude that the majority of TSS occur in the first 10% of the gene length, which is pretty much what we would expect."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
