{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "# Introduction to Gene Expression Analysis by Arrays 2 - Prelab\n",
    "\n",
    "**For this prelab, read the example code and descriptions below, and answer the questions throughout.**\n",
    "\n",
    "### I.1. Overview and Objectives\n",
    "In this lab, we will introduce you to the analysis of gene expression data, as generated by microarrays. After this prelab and some in-class activities, you will:\n",
    "\n",
    "* Learn about Normalization: what and why and an R function to normalize data: `rma()`\n",
    "* Build a design module for the analysis of your data: `model.matrix()`\n",
    "* Perform differential expression analysis: `eBayes()`\n",
    "* Report p-values from the top results: `topTable()`\n",
    "* Output results from your expression analysis: `write.table()`\n",
    "\n",
    "In the following prelab, we're going to walk through a range of commands, picking up where we left off in the previous class."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "### II.1. Analysis Pipeline: Normalization\n",
    "\n",
    "As we can see from the histogram or box plots from the previous class, the distribution of intensity values differs across each CEL file. This is problematic, because we want intensity data to be comparable across different arrays (and across treatment conditions). To deal with this issue, we need to _normalize_ the data in some way. As it turns out, there are many ways to normalize data, and many packages exist that implement a wide range of normalization procedures. \n",
    "\n",
    "One robust way to normalize data is using the Robust Multichip Average (RMA) approach. To perform normalization using RMA, simply use:\n",
    "\n",
    "    genenorm <- rma(affyRaw)\n",
    "\n",
    "These look much better, don't they?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "![example-boxplot-norm](img/ex_boxplot_norm.png \"Example Box Plot for Probe Intensity Data (Normalized)\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "![example-hist-norm](img/ex_hist_norm.png \"Example Histogram for Probe Intensity Data (Normalized)\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "### II.2. Analysis Pipeline: Design Matrix for Analysis\n",
    "\n",
    "Next up, we need to create a matrix that tells R what groups to compare against. In this example, we will design a simple, two group comparison (Liver vs. Spleen).\n",
    "    \n",
    "    group <- factor(t(pData(phenoData)[, 1]))\n",
    "    group\n",
    "\n",
    "\n",
    "    design <- model.matrix(~group, pData(phenoData))\n",
    "\n",
    "> \\> design <br \\>\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;(Intercept) GroupSpleen <br \\>\n",
    "MoGene-2_1_GA_Liver_1_C01.ga.cel&emsp;&emsp;&emsp;&emsp;&emsp;1&emsp;&emsp;&emsp;&emsp;&emsp;0 <br \\>\n",
    "MoGene-2_1_GA_Liver_A01.ga.cel&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;1&emsp;&emsp;&emsp;&emsp;&emsp;0 <br \\>\n",
    "MoGene-2_1_GA_Liver_C01.ga.cel&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;1&emsp;&emsp;&emsp;&emsp;&emsp;0 <br \\>\n",
    "MoGene-2_1_GA_Liver_G01.ga.cel&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;1&emsp;&emsp;&emsp;&emsp;&emsp;0 <br \\>\n",
    "MoGene-2_1_GA_Spleen_A01.ga.cel&emsp;&emsp;&emsp;&emsp;&emsp;1&emsp;&emsp;&emsp;&emsp;&emsp;1 <br \\>\n",
    "MoGene-2_1_GA_Spleen_C01.ga.cel&emsp;&emsp;&emsp;&emsp;&emsp;1&emsp;&emsp;&emsp;&emsp;&emsp;1 <br \\>\n",
    "MoGene-2_1_GA_Spleen_E01.ga.cel&emsp;&emsp;&emsp;&emsp;&emsp;1&emsp;&emsp;&emsp;&emsp;&emsp;1 <br \\>\n",
    "MoGene-2_1_GA_Spleen_G01.ga.cel&emsp;&emsp;&emsp;&emsp;&emsp;1&emsp;&emsp;&emsp;&emsp;&emsp;1 <br \\>\n",
    "attr(,\"assign\") <br \\>\n",
    "[1] 0 1 <br \\>\n",
    "attr(,\"contrasts\") <br \\>\n",
    "attr(,\"contrasts\")$Group <br \\>\n",
    "[1] \"contr.treatment\" <br \\>\n",
    "\n",
    "As you can see, all of the CEL files that are Liver are set to \"0\", and Spleen \"1\", and the appropriate object has been created. Here, we will measure expression in Spleen (=1), relative to Liver (=0)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "#### Q: How many samples are being compared in each group?"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {
    "collapsed": false
   },
   "source": [
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "### II.3. Analysis Pipeline: Performing differential expression analysis and summarizing results\n",
    "\n",
    "Now that we have our design matrix in hand, we can now prepare to test our probes for differences in gene expression. To perform this analysis, we will use linear regression to determine if the intensity of the probe is correlated with the effect the treatment (in this case, tissue), and if so, estimate the effect of the probe intensity (Î²). Because the number of arrays we typically analyze is small, we need to use _moderated_ t-statistics and an empirical Bayes procedure, equivalent to shrinkage of the estimated sample variances of the probe intensities towards a pooled estimate, resulting in better stability for inference (see [Smith 2004](http://www.statsci.org/smyth/pubs/ebayes.pdf) for more details).\n",
    "\n",
    "    fit <- lmFit(genenorm, design)\n",
    "    efit <- eBayes(fit)\n",
    "    \n",
    "The result is a large table of scores! We can use another function, topTable(), to obtain results\n",
    "\n",
    "    tt <- topTable(efit, sort = \"P\", n = 500)\n",
    "\n",
    "Here, I have opted to sort the Table by P-value, and am only recording the top 500 results in the scan. If I only wanted a table with all the results, I could set `n=Inf` instead, but keep in mind that will generate a VERY large file (>200 Mb).\n",
    "    \n",
    "> \\> tt[1:5, ] <br \\>\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;logFC  AveExpr&emsp;&emsp;&emsp;&emsp;&emsp;t&emsp;&emsp;&emsp;P.Value&emsp;&emsp;&ensp;adj.P.Val&emsp;&emsp;&emsp;&ensp;B <br \\>\n",
    "17405908 -4.136117 6.911107 -67.54411 5.597037e-13 1.415179e-08 20.03929 <br \\>\n",
    "17284936 -4.116741 6.375286 -63.26197 9.795902e-13 1.415179e-08 19.60969 <br \\>\n",
    "17419622 -3.586706 7.065164 -61.67103 1.217751e-12 1.415179e-08 19.43799 <br \\>\n",
    "17472364 -4.903165 5.905147 -59.14892 1.739743e-12 1.415179e-08 19.15110 <br \\>\n",
    "17219356 -3.681500 7.225212 -58.70824 1.854509e-12 1.415179e-08 19.09902 <br \\>\n",
    "\n",
    "Not surprisingly, there are several probes which are differently expressed between liver and spleen, in this case, expressed lower in _Spleen_ relative to Liver (note the negative values here). Positive scores here would indicate higher expression in Spleen, relative to Liver.\n",
    "\n",
    "**IMPORTANT NOTE: Capital letters matter for the function names here! eBayes() is not the same as ebayes(); topTable() also is not the same as toptable().** \n",
    "\n",
    "**BE SURE YOU ARE USING topTable() and eBayes() !!!!**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "#### Q: Given that ~20,000 genes tested, do any of these probes appear differentially expressed between Liver and Spleen? Why or why not?"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {
    "collapsed": false
   },
   "source": [
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "### II.4. Analysis Pipeline: Output results to file\n",
    "\n",
    "Often, we will want to output our results to a file, so that we can summarize, save results, and extract information more readily using other tools. We can use `write.table` for this:\n",
    "\n",
    "    write.table(tt, file = \"example_allresults.txt\", quote = F, sep = \",\", row.names = T, col.names = NA);\n",
    "    \n",
    "Note that adding `col.names = NA` prevents the header in the file from being \"off-by-one\" column. e.g. the first column is not logFC, but the probeID.\n",
    "\n",
    "Also note here that in this case, we're separating the results by comma (`sep = \",\"`), which generally should be fine. In some cases, you may have a table that contains not just numbers, but also words or descriptions (which may themselves include whitespaces, or even commas!).\n",
    "\n",
    "If that is the case, you can set `quote = T`, which will keep those units together. (parsing those data files in UNIX can be a trick issue, however)."
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {
  },
  "kernelspec": {
   "display_name": "R (SageMath)",
   "language": "r",
   "name": "ir-sage"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  },
  "name": "R-Bioconductor-GEO-prelab.ipynb"
 },
 "nbformat": 4,
 "nbformat_minor": 0
}